src/Wait.hs

```diff
@@ -53,23 +55,36 @@ listContentsInside test dir = do
 --   starting from when 'waitFiles' was initially called.
 --
 --   Returns a message about why you are continuing (usually a file name).
-waitFiles :: Waiter -> IO ([FilePath] -> IO [String])
+waitFiles :: forall a.  Waiter -> IO ([(FilePath, a)] -> IO (Either String [(FilePath, a)]))
 waitFiles waiter = do
     base <- getCurrentTime
-    return $ \files -> handle (\(e :: IOError) -> do sleep 1.0; return ["Error when waiting, if this happens repeatedly, raise a ghcid bug.",show e]) $ do
-        whenLoud $ outStrLn $ "%WAITING: " ++ unwords files
+    return $ \files -> handle onError (go base files)
+ where
+    onError :: IOError -> IO (Either String [(FilePath, a)])
+    onError e = sleep 1.0 >> pure (Left (show e))
+
+    go :: UTCTime -> [(FilePath, a)] -> IO (Either String [(FilePath, a)])
+    go base files = do
+        whenLoud $ outStrLn $ "%WAITING: " ++ unwords (map fst files)
         -- As listContentsInside returns directories, we are waiting on them explicitly and so
         -- will pick up new files, as creating a new file changes the containing directory's modtime.
-        files <- fmap concat $ forM files $ \file ->
-            ifM (doesDirectoryExist file) (listContentsInside (return . not . isPrefixOf "." . takeFileName) file) (return [file])
+        files <- fmap concat $ forM files $ \(file, a) ->
+            ifM
+              (doesDirectoryExist file)
+              (fmap (,a) <$> listContentsInside (return . not . isPrefixOf "." . takeFileName) file)
+              (return [(file, a)])
         case waiter of
             WaiterPoll t -> return ()
             WaiterNotify manager kick mp -> do
-                dirs <- fmap Set.fromList $ mapM canonicalizePathSafe $ nubOrd $ map takeDirectory files
+                dirs <-
+                  fmap Map.fromList
+                    $ mapM canonicalizePathSafe
+                    $ nubOrdOn fst
```

Comment:
This looks like a bug. If you have `--restart=foo --reload=foo` then the `nubOrdOn` is going to throw away one of them, whichever is second. If that's the invariant, it needs to be documented. If you think that isn't a problem cause we recover it later, then why do you need the Map in the first place?

Reply:
`fire` already explicitly lets restarts have priority over reloads, so I think it'd be neater to just not throw it away here at all.

Reply:
Thanks for catching!